File: c++/LongestCommonPrefix.cpp
class Solution
{
  public:
    string longestCommonPrefix(vector<string> &strs)
    {
        if (strs.size() == 0)
        {
            string prefix = "";
            return prefix;
        }
        return longestPrefix(strs, 0, strs.size() - 1);
    }

  private:
    string longestPrefix(vector<string> &strs, int l, int r)
    {
        if (l == r)
        {
            return strs[l];
        }
        else
        {
            int mid = (l + r) / 2;
            string lcpLeft = longestPrefix(strs, l, mid);
            string lcpRight = longestPrefix(strs, mid + 1, r);
            return commonPrefix(lcpLeft, lcpRight);
        }
    }

    string commonPrefix(string lcpLeft, string lcpRight)
    {
        int min_size = min(lcpLeft.size(), lcpRight.size());
        for (int i = 0; i < min_size; i++)
        {
            if (lcpLeft[i] != lcpRight[i])
            {
                return lcpLeft.substr(0, i);
            }
        }
        return lcpLeft.substr(0, min_size);
    }
};

File: python/longestCommonPrefix.py
class Solution:
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        prefix = ''
        n = len(strs)
        if n == 0:
            return prefix

        str1 = strs[0]
        idx = len(str1)
        for i in range(1, n):
            s = strs[i]
            j = 0
            while j <= idx and j < len(str1) and j < len(s):
                if str1[j] != s[j]:
                    idx = min(idx, j)
                    break
                j += 1
            idx = min(idx, j)

        return str1[:idx]
